### PR TITLE
Subqueries on RealmLists now use the proper tablespec 

### PR DESCRIPTION
**Pull request by cmelchior**
This is a fix for: https://github.com/realm/realm-java/issues/707

Also improved error reporting for mismatching types when doing queries.

@bmunkholm @kneth @emanuelez

### CODE REVIEW

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 1]
```diff
@@ -63,26 +62,14 @@ public RealmQuery(Realm realm, Class<E> clazz) {
...
-
     /**
      * Create a RealmQuery instance from a @{link io.realm.RealmResults}.
+     *
      * @param realmList   The @{link io.realm.RealmResults} to query
      * @param clazz       The class to query
      * @throws java.lang.RuntimeException Any other error
      */
     public RealmQuery(RealmResults realmList, Class<E> clazz) {
-        this.realmList = realmList;
```
Comment (kneth) [inline realm/src/main/java/io/realm/RealmQuery.java]: I guess that this is the fix :-)

**Comment by kneth** [comment]
:+1:

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 2]
```diff
@@ -725,6 +726,27 @@ public void testRealmQueryContainsAndCaseSensitiveWithNonLatinCharacters() {
         assertEquals(0, resultList.size());
     }
 
+
+    public void testRealmQueryLink() {
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: This should maybe go in a new `RealmQueryTest` file.

**Comment by emanuelez** [comment]
:+1:

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 3]
```diff
@@ -94,14 +81,20 @@ public RealmQuery(RealmResults realmList, Class<E> clazz) {
...
-        if(realmList != null) {
-            return realmList.getTable();
-        } else {
-            return realm.getTable(clazz);
+    RealmQuery(Realm realm, TableQuery query, Class<E> clazz) {
+        this.realm = realm;
+        this.clazz = clazz;
+        this.query = query;
+        TableOrView dataStore = getTable();
+        for (int i = 0; i < dataStore.getColumnCount(); i++) {
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/RealmQuery.java]: Can't we use `Realm.columnIndices` ?

File: changelog.txt [hunk 4]
```diff
@@ -3,6 +3,7 @@
  * Lenient UTF-8 to UTF-16 transcoding by inserting replacement characters.
  * Adding more logging capabilities at the JNI level.
  * Added proper encryption support. NOTE: The key has been increased from 32 bytes to 64 bytes (see example).
+ * Fixed bug when doing subqueries on RealmLists.
```
Comment (bmunkholm) [inline changelog.txt]: Can we detail how people would experience this?

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 2 above)
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: agree.

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 5]
```diff
@@ -160,8 +153,15 @@ private boolean containsDot(String s) {
...
                 throw new IllegalArgumentException(String.format("Field '%s' does not exist.", fieldName));
             }
-            if (fieldType != table.getColumnType(columns.get(fieldName))) {
-                throw new IllegalArgumentException(String.format("Field '%s': type mismatch.", fieldName));
+
+            ColumnType tableColumnType = table.getColumnType(columns.get(fieldName));
+            if (fieldType != tableColumnType) {
+                throw new IllegalArgumentException(String.format(
+                        "Field '%s': type mismatch. Was %s, expected %s.",
+                        fieldName,
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/RealmQuery.java]: This is a new "style" to split parameters with one new on each line. Seems very verbose.

**Comment by bmunkholm** [comment]
otherwise +1

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 6]
```diff
@@ -1131,7 +1120,7 @@ public long count() {
     public RealmResults<E> findAll(String fieldName, boolean sortAscending) {
         TableView tableView = query.findAll();
         TableView.Order order = sortAscending ? TableView.Order.ascending : TableView.Order.descending;
-        Integer columnIndex = columns.get(fieldName);
+        Long columnIndex = columns.get(fieldName);
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/RealmQuery.java]: Why boxed Long here?

Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmQuery.java]: Because there is a null check just below it. It is used to indicate not present.

**Pull request merged**